### PR TITLE
layout: Dramatically improve performance of layout queries and `requestAnimationFrame()`.

### PR DESCRIPTION
**Pull request by pcwalton**
See individual commits for details. This improves nytimes.com CPU usage and gets it within the ballpark of Firefox.

r? @emilio

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14442)

### CODE REVIEW

**Comment by fitzgen** [comment]
Regarding ba274c9: Is an RAF that performs canvas draws but no dom mutations considered "spurious"? Is that desirable?

**Comment by pcwalton** [comment]
@fitzgen No, it shouldn't be considered spurious. Does it dirty the canvas? If not, I'm not sure what the best way to detect this is.

**Comment by pcwalton** [comment]
The canvas element is dirtied when you draw to it, so there should be no problems here.

**Review by emilio — COMMENTED** [review]
r=me on the layout changes, they're great :)

I'm not sure we can fake animation frame callbacks using a timer per spec. IIRC we should run them at a very specific timing per [the processing model](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model), and we're already awful at it unfortunately.

I'd love to see this, and it may be the case that results are black-box identical in this case (though I think not because at least the last fake callback that modifies something and makes us go back to real animation frames runs out of sync).

I'd probably defer to @jdm or @Ms2ger to approve that bit, or request more proof of that being the case.

**Comment by pcwalton** [comment]
@emilio I don't see anything in that list that mandates *when* animation frame callbacks run, only that they run at specific times relative to other events. We already don't try to do that, I don't think. Indeed it'd be impossible for the spec to mandate a specific time other than describing what a VBLANK is (and if it did, then every other browser on non-Windows platforms would fail it because they don't sync to VBLANK properly).

(On a side note, it's irritating that the spec technically forbids off main thread CSS animations, but I'm going to just break the spec because that's dumb.)

**Comment by emilio** [comment]
On Fri, Dec 02, 2016 at 11:39:39PM -0800, Patrick Walton wrote:
> We already don't try to do that, I don't think. Indeed it'd be
> impossible for the spec to mandate a specific time other than
> describing what a VBLANK is (and if it did, then every other browser
> on non-Windows platforms would fail it because they don't sync to
> VBLANK properly).

That's fair, that's why I lean towards merging this. But I think we
eventually should try to conform to it, given there are pages that rely
on it, and this will be more complexity to rewrite/remove when we want
to do it.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -3178,3 +3226,25 @@ pub enum FocusEventType {
...
+/// refresh.
+#[derive(JSTraceable, HeapSizeOf)]
+pub struct FakeRequestAnimationFrameCallback {
+    /// The document.
+    #[ignore_heap_size_of = "non-owning"]
+    document: Trusted<Document>,
+}
+
+impl FakeRequestAnimationFrameCallback {
+    #[allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/dom/document.rs]: This shouldn't be necessary.

**Review by jdm — CHANGES_REQUESTED** [review]
This looks fine to me. Sorry that it was overlooked for so long!

**Comment by nox** [comment]
@jdm Want me to carry this? @pcwalton is busy with his Pathfinder adventures AFAIK.

**Comment by jdm** [comment]
That would be great!

**Comment by nox** [comment]
I carried the first 2 commits in #15816. 👶👣

**Comment by nox** [comment]
@pcwalton Are you sure the commit "query: Stop searching the entire flow tree to resolve script-to-layout" works as intended?

**Comment by nox** [comment]
Lack of response, I'm closing this.

**Pull request closed without merging**